test/integration/controlplane/transformation/testutil/restart_server.go

```diff
@@ -0,0 +1,507 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+/*
+The etcd package can be utilized in tests that involve the handling of
+ETCD storage data. It starts an API server that can be used for tests that require
+one of every type of resource. If a test requires the restart of an API server,
+then StartRealAPIServerOrDieForKMS and Restart can be used. It can restart the API server
+without impacting the ETCD data or affecting client connections.
+*/
+package testutil
+
+import (
+	"bytes"
+	"context"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/pem"
+	"fmt"
+	"math"
+	"math/big"
+	"net"
+	"net/http"
+	"os"
+	"path/filepath"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/spf13/pflag"
+	clientv3 "go.etcd.io/etcd/client/v3"
+
+	apiextensionsclientset "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/apiserver/pkg/storage/storagebackend"
+	"k8s.io/client-go/dynamic"
+	clientset "k8s.io/client-go/kubernetes"
+	restclient "k8s.io/client-go/rest"
+	"k8s.io/client-go/util/cert"
+	utiltesting "k8s.io/client-go/util/testing"
+	"k8s.io/klog/v2"
+	"k8s.io/kubernetes/cmd/kube-apiserver/app"
+	"k8s.io/kubernetes/cmd/kube-apiserver/app/options"
+	_ "k8s.io/kubernetes/pkg/controlplane"
+	"k8s.io/kubernetes/test/integration"
+	"k8s.io/kubernetes/test/integration/etcd"
+	"k8s.io/kubernetes/test/integration/framework"
+	"k8s.io/kubernetes/test/utils"
+	netutils "k8s.io/utils/net"
+)
+
+// This key is for testing purposes only and is not considered secure.
+const ecdsaPrivateKey = `-----BEGIN EC PRIVATE KEY-----
+MHcCAQEEIEZmTmUhuanLjPA2CLquXivuwBDHTt5XYwgIr/kA1LtRoAoGCCqGSM49
+AwEHoUQDQgAEH6cuzP8XuD5wal6wf9M6xDljTOPLX2i8uIp/C/ASqiIGUeeKQtX0
+/IR3qCXyThP/dbCiHrF3v1cuhBOHY8CLVg==
+-----END EC PRIVATE KEY-----`
+
+// APIServer represents a running API server that is ready for use
+// The Cleanup func must be deferred to prevent resource leaks
+type APIServer struct {
```

Comment:
I'm having a hard time following why we need a dedicated helper for handling server restarts. We have other integration tests that start multiple servers with different config / feature enablement against the same etcd data:

https://github.com/kubernetes/kubernetes/blob/eee56d56edca5d618ab004b2051f522de75b4573/test/integration/apiserver/admissionwebhook/broken_webhook_test.go#L48-L83

https://github.com/kubernetes/kubernetes/blob/eee56d56edca5d618ab004b2051f522de75b4573/test/integration/auth/rbac_test.go#L743-L797

https://github.com/kubernetes/kubernetes/blob/eee56d56edca5d618ab004b2051f522de75b4573/test/integration/storageversion/storage_version_filter_test.go#L150-L182

Is there a reason we can't do that directly in the test?

Reply:
@liggitt Yes we started with using the same testserver all these tests use. However a test that does a GET on a secret that was created BEFORE a restart would fail as the tearDown function removes the storage layer and the tempdir that etcd uses.
https://github.com/kubernetes/kubernetes/blob/e3bc35bc1bc7b399b564f2c4efc75eb9959e70f7/cmd/kube-apiserver/app/testing/testserver.go#L125-L128

https://github.com/kubernetes/kubernetes/blob/e3bc35bc1bc7b399b564f2c4efc75eb9959e70f7/cmd/kube-apiserver/app/testing/testserver.go#L138

Reply:
and we did not want to update the existing testserver file as it can impact other tests.

Reply:
etcd data shouldn't be located in that tmpdir. `storageConfig` is passed in, which means it already has a location for data. `result.TmpDir, err = os.MkdirTemp("", "kubernetes-kube-apiserver")` is where other stuff (serving certs, etc) get located. The examples I linked to bring up and tear down multiple servers talking to the same etcd with storage persisting between them... that should be possible